packages/data-context/src/data/ProjectConfigManager.ts

```diff
@@ -0,0 +1,790 @@
+
+    for (const handler of this.options.handlers) {
+      handler(ipc)
+    }
+
+    const { promise } = this.registerSetupIpcHandlers(ipc)
+
+    const overrides = config[this._testingType] ?? {}
+    const mergedConfig = { ...config, ...overrides }
+
+    // alphabetize config by keys
+    let orderedConfig = {} as Cypress.PluginConfigOptions
+
+    Object.keys(mergedConfig).sort().forEach((key) => {
+      const k = key as keyof typeof mergedConfig
+
+      // @ts-ignore
+      orderedConfig[k] = mergedConfig[k]
+    })
+
+    ipc.send('setupTestingType', this._testingType, {
+      ...orderedConfig,
+      projectRoot: this.options.projectRoot,
+      configFile: this.configFilePath,
+      version: this.options.cypressVersion,
+      testingType: this._testingType,
+    })
+
+    return promise
+  }
+
+  private async handleSetupTestingTypeReply (ipc: ProjectConfigIpc, loadConfigReply: LoadConfigReply, result: SetupNodeEventsReply) {
+    this._registeredEvents = {}
+
+    for (const { event, eventId } of result.registrations) {
+      debug('register plugins process event', event, 'with id', eventId)
+
+      this.registerEvent(event, function (...args: any[]) {
+        return new Promise((resolve, reject) => {
+          const invocationId = _.uniqueId('inv')
+
+          debug('call event', event, 'for invocation id', invocationId)
+
+          ipc.once(`promise:fulfilled:${invocationId}`, (err: any, value: any) => {
+            if (err) {
+              debug('promise rejected for id %s %o', invocationId, ':', err.stack)
+              reject(_.extend(new Error(err.message), err))
+
+              return
+            }
+
+            if (value === UNDEFINED_SERIALIZED) {
+              value = undefined
+            }
+
+            debug(`promise resolved for id '${invocationId}' with value`, value)
+
+            return resolve(value)
+          })
+
+          const ids = { invocationId, eventId }
+
+          // no argument is passed for cy.task()
+          // This is necessary because undefined becomes null when it is sent through ipc.
+          if (event === 'task' && args[1] === undefined) {
+            args[1] = {
+              __cypress_task_no_argument__: true,
+            }
+          }
+
+          ipc.send('execute:plugins', event, ids, args)
+        })
+      })
+    }
+
+    const cypressEnv = await this.loadCypressEnvFile()
+    const fullConfig = await this.buildBaseFullConfig(loadConfigReply.initialConfig, cypressEnv, this.options.modeOptions)
+
+    const finalConfig = this._cachedFullConfig = this.options.updateWithPluginValues(fullConfig, result.setupConfig ?? {})
+
+    // This happens automatically with openProjectCreate in run mode
+    if (!this.options.isRunMode) {
+      if (!this._initializedProject) {
+        this._initializedProject = await this.options.initializeActiveProject({})
+      } else {
+        // TODO: modify the _initializedProject
+      }
+    }
```

Comment:
I think this will probably live outside of the config manager, assuming that we're able to handle changes to the config without destroying the `initializedProject`, since that's what holds the server reference.